python/pyarrow/dataset.py

```diff
@@ -232,104 +228,261 @@ def _ensure_format(obj):
+    if file_info is None:
+        file_info = filesystem.get_file_info([path])[0]
+
+    # depending on the path type either return with a recursive
+    # directory selector or as a list containing a single file
+    if file_info.type == FileType.Directory:
+        paths_or_selector = FileSelector(path, recursive=True)
+    elif file_info.type == FileType.File:
+        paths_or_selector = [path]
+    else:
+        raise FileNotFoundError(path)
+
+    return filesystem, paths_or_selector
 
-def dataset(paths_or_factories, filesystem=None, partitioning=None,
-            format=None, schema=None):
+
+def _filesystem_dataset(source, schema=None, filesystem=None,
+                        partitioning=None, format=None,
+                        partition_base_dir=None, exclude_invalid_files=None,
+                        ignore_prefixes=None):
+    """
+    Create a FileSystemDataset which can be used to build a Dataset.
+
+    Parameters are documented in the dataset function.
+
+    Returns
+    -------
+    FileSystemDataset
+    """
+    format = _ensure_format(format or 'parquet')
+    partitioning = _ensure_partitioning(partitioning)
+
+    if isinstance(source, (list, tuple)):
+        fs, paths_or_selector = _ensure_multiple_sources(source, filesystem)
+    else:
+        fs, paths_or_selector = _ensure_single_source(source, filesystem)
+
+    options = FileSystemFactoryOptions(
+        partitioning=partitioning,
+        partition_base_dir=partition_base_dir,
+        exclude_invalid_files=exclude_invalid_files,
+        ignore_prefixes=ignore_prefixes
+    )
+    factory = FileSystemDatasetFactory(fs, paths_or_selector, format, options)
+
+    return factory.finish(schema)
+
+
+def _union_dataset(source, schema=None, **kwargs):
+    if any(v is not None for v in kwargs.values()):
+        raise ValueError(
+            "When passing a list of Datasets, you cannot pass any additional "
+            "arguments"
+        )
+
+    if schema is None:
+        # unify the children datasets' schemas
+        schema = pa.unify_schemas([ds.schema for ds in source])
+
+    # create datasets with the requested schema
+    children = [ds.replace_schema(schema) for ds in source]
+
+    return UnionDataset(schema, children)
+
+
+def dataset(source, schema=None, format=None, filesystem=None,
+            partitioning=None, partition_base_dir=None,
+            exclude_invalid_files=None, ignore_prefixes=None):
     """
     Open a dataset.
 
     Parameters
     ----------
-    paths_or_factories : path or list of paths or factory or list of factories
+    source : path, list of paths, dataset or list of datasets
         Path to a file or to a directory containing the data files, or a list
         of paths for a multi-directory dataset. To have more control, a list of
         factories can be passed, created with the ``factory()`` function (in
         this case, the additional keywords will be ignored).
```

Comment:
mention the list-of-datasets argument case? In general, this section seems like it might be clearer if presented as a definition list keyed by type(source)

Reply:
Updated with some additional docstrings and examples (shamelessly copied some parts from Joris' [guide](https://github.com/apache/arrow/pull/6779))